statsmodels/genmod/generalized_linear_model.py

```diff
@@ -700,18 +713,82 @@ def estimate_tweedie_power(self, mu, method='brentq', low=1.01, high=5.):
         power : float
             The estimated shape or power
         """
+        if exposure is not None and not isinstance(self.family.link,
+                                                   families.links.Log):
+            raise ValueError("exposure can only be used with the log link function")
+
+        # Use fit exposure if appropriate
+        if exposure is None and hasattr(self, 'exposure'):
+            # Already logged
+            exposure = np.exp(self.exposure)
+        elif exposure is None:
+            exposure = np.ones(len(self.endog))
+
+        # Use fit freq_weights if appropriate
+        if freq_weights is None and hasattr(self, 'freq_weights'):
+            # Already logged
+            freq_weights = self.freq_weights
+        elif exposure is None:
+            freq_weights = np.ones(len(self.endog))
+
+        weights = freq_weights * exposure
+
+        endog2 = self.endog / exposure
+        mu2 = mu / exposure
+
         if method == 'brentq':
             from scipy.optimize import brentq
 
-            def psi_p(power, mu):
-                scale = ((self.freq_weights * (self.endog - mu) ** 2 /
+            low = kwargs.get('low', 1)
+            high = kwargs.get('high', 10)
+
+            def x2scale(power):
+                scale = ((freq_weights * (self.endog - mu) ** 2 /
                           (mu ** power)).sum() / self.df_resid)
-                return (np.sum(self.freq_weights * ((self.endog - mu) ** 2 /
-                               (scale * (mu ** power)) - 1) *
-                               np.log(mu)) / self.freq_weights.sum())
-            power = brentq(psi_p, low, high, args=(mu))
+                return scale
+
+            def psi_p(power, endog2, mu2, weights):
+                scale = x2scale(power)
+                p = (np.sum(weights * ((endog2 - mu2) ** 2 /
+                            (scale * (mu2 ** power)) - 1) * np.log(mu2)) /
+                     weights.sum())
+                return p
+
+            power = brentq(psi_p, low, high, args=(endog2, mu2, weights))
+            if both:
+                scale = x2scale(power)
+                return scale, power
+            else:
+                return power
+
+        elif method == 'perry':
+            y = np.power(endog2 - mu2, 2)
+            x = np.column_stack((np.ones(len(endog2)), np.log(mu2)))
+            model = GLM(y, x, freq_weights=weights,
+                        family=families.Gamma(families.links.log))
+            res = model.fit()
+            scale, power = res.params
+
+            if both:
+                return np.exp(scale), power
+            else:
+                return power
+
+        elif method == 'taylor':
+            from statsmodels.regression.linear_model import WLS
+
+            y = np.log(np.power(endog2 - mu2, 2))
```

Comment:
my guess is that we need to clip `endog2 - mu2` to be larger than zero. Somebody will come up with a case with zero residuals. I'm not sure what's a good threshold, 1e-6?